### PR TITLE
PUBDEV-5789: cleaner message after failure to locally start server

### PR DESCRIPTION
**Pull request by pkozelka**
- removed useless call to curl
- replaced with message suggesting that the user should check Java version used for starting h2o
 
When the failure occurs, this can be for couple of reasons, including (1) connection problems and (2) problems to start the application. Before this change, the console output always showed the symptom, twice (failure to connect) but did not indicate possible causes.

In the first case, the symptom is already indicated by the message from R output (which anyway seems to be equal to the one from `curl`). Therefore calling curl again (probably meant just to get the error message) does not seem useful.

In the latter case, one of possible causes is, incorrect version of Java. So the user is instructed to check it.
For any other causes, the user can explore the indicated console log files; and here it comes handy to get rid of the (IMHO useless) curl output which would otherwise place the log pointer several lines more behind in the console history.

Now the error output in RStudio looks like this:
```
...
> h2o.init()
Reading in config file: ./.h2oconfig

H2O is not running yet, starting it now...

Note:  In case of errors look at the following log files:
    /tmp/RtmpssoTKx/h2o_pk_started_from_r.out
    /tmp/RtmpssoTKx/h2o_pk_started_from_r.err

java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) 64-Bit Server VM (build 20.45-b01, mixed mode)

Starting H2O JVM and connecting: ............................................................ 
[1] "localhost"
[1] 54321
[1] TRUE
[1] -1
[1] "Failed to connect to localhost port 54321: Connection refused"
[1] "Please check JAVA version used for running the H2O server"
Error in h2o.init() : H2O failed to start, stopping execution.
```

BTW, the original `curl` command was also somewhat ill; it always assumes default values for `ip` and `port` parameters. If it should stay, its params should be rather constructed from actual parameter values.

### CODE REVIEW

**Comment by Pscheidl** [comment]
Actually we discussed @michalkurka 's proposal yesterday already.

Looks like some users may get confused a little bit how to set-up Java in their environment properly. This is a perfect example: https://groups.google.com/forum/#!topic/h2ostream/3vBMiIOhOVs

## Documentation issues
- The documentation of init() method says nothing about recommended Java version or Java version issues in general. We should consider adding a note about that in there. Maybe double-check with @angela0xdata (Angela Bartz) once we decide ? A little bit more documentation for the end user wouldn't hurt.
- Documentation changes should of course be reflected here: http://docs.h2o.ai/h2o/latest-stable/h2o-docs/starting-h2o.html#from-r
- The error message could point to our H2O online documentation (not to the CRAN PDF), where the user is able to get tips on how to solve that. At least tell them to check their PATH and JAVA_HOME. And give them a list of supported version. People won't look this information up in other sections of the documentation (my opinion).

## To check or not to check `java -version`
- I don't like the `Failed to connect to localhost port 54321: Connection refused` message followed by a message about checking Java version. This is not confusing for us developers (maybe), but it is definitely confusing to data scientists. 

My proposed solution in such case would be:
1. Invoke `java -version`. Try to extract version number. If unsuccessfull, print a message about this check being unsuccessful: "Java version check unsuccessful. Attempting to start H2O."
2. If the version check is successful, start H2O if version is from the group of supported version.
3. If the version check is unsuccessful, do not attempt to start H2O and print a message about that. With documentation link. As demonstrated above, users may have multiple Java installations on their machine and we want them to know there is a possibility of setting the right one into environment variables.

@pkozelka  also proposed a solution with custom class compiled against old bytecode versions, that just prints the Java version. H2O R could run it quickly. This would avoid us relying on the message format `java -version` gives us. I do think this solution is elegant, yet very time consuming and the solution proposed above follows the 80/20 rule. It is very unlikely for the message to change, as we don't support platforms like IcedTea and other projects (does that even exist anymore ?).

- The check is not really duplicated. In this scenario, it is. We need to keep the runtime check, as there are many ways to run H2O, including Apache Spark, running it standalone etc. etc...
- I do agree with @pkozelka on adding a notice to the original check in H2O class for other people to check R/Python client code as they change that piece of code.

**Comment by angela0xdata** [comment]
If a documentation change for init() is added to R, then it should be added to the "From Python" section in the user guide and to the R and Python docs, too. Should be a separate Jira for this change. Also, just a quick note that Java version is listed in the requirements section of the User Guide: http://docs.h2o.ai/h2o/latest-stable/h2o-docs/welcome.html#requirements

**Comment by michalkurka** [comment]
@pkozelka @Pscheidl The current behavior is difficult to decipher for the users. Please refer to see the output of a failed attempt to start the server on Java 6: https://0xdata.atlassian.net/browse/PUBDEV-5789 - a regular user has no idea what is wrong.

Checking the same thing on the client and on backend doesn't raise concerns to me. It is analogous to validating input both on the frontend and on the backend of a web application (or really any client-server scenario).

Another thing is that the check you are refering to will not even run because the jar is compiled with target java version 1.7.

In my opinion we should adopt best-effort approach and prevent the most common scenarios. Users are usually using openjdk or oracle jdk. We can incorporate a check that will reject versions that we _know_ are no compatible and make most people happy.

Just like in the frontend validation example - this is a convenience measure not a fail-safe mechanism.

**Comment by pkozelka** [comment]
OK so I added the pre-start check which implements the same logic as `H2O.checkUnsupportedJava`.
And comments to simplify future maintenance.

If you believe this is OK, I would continue by enhancing the documentation with list of supported versions, so that I can add the doc link to the error messages.

Here is how the output looks now, when java is set to 1.6:

```
> h2o.init()

H2O is not running yet, starting it now...
Error in h2o.init() : 
  Only Java 1.7-1.8, 9 and 10 is supported, your version is 1.6.0_45
```

And when there is no java:

```
> h2o.init()

H2O is not running yet, starting it now...
Error in .h2o.checkJava() : 
  Cannot find Java. Please install the latest JRE from
http://www.oracle.com/technetwork/java/javase/downloads/index.html
```

**Comment by pkozelka** [comment]
@angela0xdata just noticed your comments - good that we already do have required java version specified.
However, it might be worth mentioning that we support versions "up to 10" besides current "7 and later", right?

**Comment by Pscheidl** [comment]
@pkozelka Java 11 is not out yet.

**Comment by michalkurka** [comment]
> OK so I added the pre-start check which implements the same logic as H2O.checkUnsupportedJava.

Let me clarify. I don't think the logic should be the same. Instead of _allowing_ to run on supported versions (like in Java), we should be _rejecteting_ known unsupported versions in R.

We will be likely adding support for J11 once it is released but we will be unlikely deprecating support for J7 anytime soon.

Error on the side that doesn't harm users if we ourselves make a mistake (eg. we can update Java check but forget to update the R client check).

File: h2o-r/h2o-package/R/connection.R [hunk 1]
```diff
@@ -163,6 +163,14 @@ h2o.init <- function(ip = "localhost", port = 54321, startH2O = TRUE, forceDL =
       stop("Cannot connect to H2O server. Please check that H2O is running at ", h2o.getBaseURL(tmpConn))
     else if (ip == "localhost" || ip == "127.0.0.1") {
       cat("\nH2O is not running yet, starting it now...\n")
+      command <- .h2o.checkJava()
+      javaVersionOutput <- system2(command, "-version", stdout=TRUE, stderr=TRUE)
+      v <- stringr::str_replace(javaVersionOutput[1], "^java version \"(.*)\"$","\\1")
+
+      # NOTE for developers: keep the following heuristic in sync with java code - see file connection.R near line 1850
+      if (!any(grepl("^1\\.7\\.",v), grepl("^1\\.8\\.",v), grepl("^9\\.", grepl("^10\\.", v)))) {
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/connection.R]: This is correct. However, I think it would be better to reuse the functionality from line 527 and move the logic there. You attempt to call `java -version` another time here. This additional call is in vain. They store the result of the call and then they run checks against it. Why wouldn't we reuse the same mechanism ?

``` plain
  jver <- tryCatch({system2(command, "-version", stdout = TRUE, stderr = TRUE)}, # RESULT STORED HERE !!!!
      error = function(err) {
        print(err)
        stop("You have a 32-bit version of Java. H2O works best with 64-bit Java.\n",
        "Please download the latest Java SE JDK 8 from the following URL:\n",
        "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
      }
    )

  if(any(grepl("GNU libgcj", jver))) { # CHECKED AGAINST ALREADY FETCHED OUTPUT !!!
    stop("Sorry, GNU Java is not supported for H2O.\n",
         "Please download the latest Java SE JDK 8 from the following URL:\n",
         "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
  }
```

File: h2o-r/h2o-package/R/connection.R [hunk 2]
```diff
@@ -190,8 +198,7 @@ h2o.init <- function(ip = "localhost", port = 54321, startH2O = TRUE, forceDL =
         print(rv$httpStatusCode)
         print(rv$curlErrorMessage)
 
-        #try a hail mary curl
-        print(system("curl 'http://localhost:54321'"))
+        print("Please check JAVA version used for running the H2O server")
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/connection.R]: Why do we need additional check here ?

Comment (pkozelka) [reply h2o-r/h2o-package/R/connection.R]: now we don't, it's gone now, thanks

File: h2o-r/h2o-package/R/connection.R (hunk 1 above)
Comment (pkozelka) [reply h2o-r/h2o-package/R/connection.R]: true, I will change it this way

**Comment by pkozelka** [comment]
Now I moved the check to proper place, as @Pscheidl suggested.
@michalkurka as you wanted, I also changed the version checking logic in R to blacklisting unsupported java versions (which currently includes version1.2-1.6)

With the latter change, I think there is probably no need to make any docs changes; the text is sufficient and based on our discussions, we don't need to be very explicit about version 10 being the latest supported one. It makes no troubles until 11 is out, and perhaps not even then.

**Comment by Pscheidl** [comment]
Wait for @michalkurka before merging. 

Also,we have a release branch named rel-wright. There should be another PR against that. I did not mention that to @pkozelka - let me explain via Slack.

File: h2o-core/src/main/java/water/H2O.java [hunk 3]
```diff
@@ -1845,7 +1845,8 @@ public static void gc() {
    */
   public static boolean checkUnsupportedJava() {
     String version = System.getProperty("java.version");
-    if (version != null && !(version.startsWith("1.7") || version.startsWith("1.8") || version.startsWith("9") || version.startsWith("10"))) {
+    // NOTE for developers: make sure that the following whitelist is logically consistent with whitelist in R code - see file connection.R near line 536
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/H2O.java]: Unnecessary change, should be tested. (A simple unit test would do).

File: h2o-r/h2o-package/R/connection.R [hunk 4]
```diff
@@ -534,6 +532,12 @@ h2o.clusterStatus <- function() {
       }
     )
 
+  # NOTE for developers: keep the following blacklist in logically consistent with whitelist in java code - see water.H2O.checkUnsupportedJava, near line 1849
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/connection.R]: May be tested as well.

**Comment by pkozelka** [comment]
corresponding PR for release branch is #2744

File: h2o-r/h2o-package/R/connection.R [hunk 5]
```diff
@@ -533,13 +546,12 @@ h2o.clusterStatus <- function() {
         "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
       }
     )
-
-  if(any(grepl("GNU libgcj", jver))) {
-    stop("Sorry, GNU Java is not supported for H2O.\n",
-         "Please download the latest Java SE JDK 8 from the following URL:\n",
-         "http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html")
+  jver_error <- .h2o.check_java_version(jver);
+  if (!is.null(jver_error)) {
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/connection.R]: This is a good refactoring.

File: h2o-r/h2o-package/R/connection.R [hunk 6]
```diff
@@ -498,6 +496,21 @@ h2o.clusterStatus <- function() {
...
+#
+.h2o.check_java_version <- function(jver = NULL) {
+  if(any(grepl("GNU libgcj", jver))) {
+    return("Sorry, GNU Java is not supported for H2O.")
+  }
+  # NOTE for developers: keep the following blacklist in logically consistent with whitelist in java code - see water.H2O.checkUnsupportedJava, near line 1849
+  if (any(grepl("^java version \"1\\.[1-6]\\.", jver))) {
+    return(paste0("Your java is not supported: ", jver[1]))
+  }
+  return(NULL)
```
Comment (Pscheidl) [inline h2o-r/h2o-package/R/connection.R]: May be just NULL. FYI. Do not change it please :)

**Comment by Pscheidl** [comment]
Nicely done. Avoided duplication of code and made everything testable.
